File: cpp/2021/sept/array-nesting.cc
#include <string>
#include <vector>

using namespace std;

class Solution {
public:
    int arrayNesting(vector<int>& nums) {
        int          res = 0, n = nums.size();
        vector<bool> seen(n);
        for (int i : nums) {
            int cnt = 0;
            while (!seen[i]) {
                seen[i] = true;
                cnt++;
                i = nums[i];
            }
            res = max(res, cnt);
        }
        return res;
    }
};

File: code/go/2021/sept/array-nesting.go
package sept

func arrayNesting(nums []int) int {
	res, n := 0, len(nums)
	seen := make([]bool, n)
	for i := range nums {
		cnt := 0
		for !seen[i] {
			seen[i] = true
			cnt++
			i = nums[i]
		}
		res = max(res, cnt)
	}
	return res
}
